core/domain/state_domain.py

```diff
@@ -916,8 +938,946 @@ def get_all_outcomes(self) -> List[Outcome]:
+                    element = rule_spec.inputs['x']
+                    position = rule_spec.inputs['y']
+                    ele_x_at_y_rules.append(
+                        {'element': element, 'position': position}
+                    )
+
+                if (
+                    rule_spec.rule_type ==
+                    'IsEqualToOrderingWithOneItemAtIncorrectPosition'
+                ):
+                    equal_ordering_one_at_incorec_posn.append(
+                        rule_spec.inputs['x']
+                    )
+
+                if rule_spec.rule_type == 'IsEqualToOrdering':
+                    # `IsEqualToOrdering` rule should not have empty values.
+                    if len(rule_spec.inputs['x']) <= 0:
+                        raise utils.ValidationError(
+                            f'The rule \'{rule_spec_index}\'of '
+                            f'answer group \'{ans_group_index}\', '
+                            f'having rule type IsEqualToOrdering '
+                            f'should not have empty values.'
+                        )
+                    # `IsEqualToOrdering` rule should always come before
+                    # `HasElementXAtPositionY` where element `X` is present
+                    # at position `Y` in `IsEqualToOrdering` rule.
+                    for ele in ele_x_at_y_rules:
+                        ele_position = ele['position']
+                        ele_element = ele['element']
+                        rule_choice = rule_spec.inputs['x'][
+                            ele_position - 1]
+
+                        for choice in rule_choice:
+                            if choice == ele_element:
+                                raise utils.ValidationError(
+                                    f'Rule - {rule_spec_index} of '
+                                    f'answer group {ans_group_index} '
+                                    f'will never be match '
+                                    f'because it is made redundant by the '
+                                    f'HasElementXAtPositionY rule above.'
+                                )
+                    # `IsEqualToOrdering` should always come before
+                    # `IsEqualToOrderingWithOneItemAtIncorrectPosition` when
+                    # they are off by one value.
+                    item_to_layer_idx = {}
+                    for layer_idx, layer in enumerate(
+                        rule_spec.inputs['x']):
+                        for item in layer:
+                            item_to_layer_idx[item] = layer_idx
+
+                    for ele in equal_ordering_one_at_incorec_posn:
+                        wrong_positions = 0
+                        for layer_idx, layer in enumerate(ele):
+                            for item in layer:
+                                if layer_idx != item_to_layer_idx[item]:
+                                    wrong_positions += 1
```

Comment:
Why do we need the count of wrong positions? Bool would be enough, right?

Reply:
Updated